File: mobile/src/main/java/info/papdt/express/helper/ui/common/AbsFragment.java
package info.papdt.express.helper.ui.common;

import android.app.Activity;
import android.app.Fragment;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import info.papdt.express.helper.support.Settings;

public abstract class AbsFragment extends Fragment {

	protected View rootView;
	private Settings mSettings;
	private Context mContext;

	protected abstract @LayoutRes int getLayoutResId();

	protected abstract void doCreateView(View rootView);

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		mSettings = Settings.getInstance(getContext());
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup parent, Bundle state) {
		if (parent != null) {
			rootView = inflater.inflate(this.getLayoutResId(), parent, false);
			mContext = parent.getContext();
		} else {
			rootView = inflater.inflate(this.getLayoutResId(), null);
		}
		this.doCreateView(rootView);

		return rootView;
	}

	@Override
	@SuppressWarnings("deprecation")
	public void onAttach(Activity activity) {
		super.onAttach(activity);

		// Use activity instead of getApplicationContext in order to prevent memory leak
		mContext = activity;
	}

	@Nullable
	protected <T extends View> T $(@IdRes int viewId) {
		return rootView != null ? (T) rootView.findViewById(viewId) : null;
	}

	@Override
	public Context getContext() {
		return mContext;
	}

	protected Settings getSettings() {
		return mSettings;
	}

}

File: mobile/src/main/kotlin/info/papdt/express/helper/ui/common/AbsFragment.kt
package info.papdt.express.helper.ui.common

import android.content.Context
import android.os.Bundle
import androidx.annotation.IdRes
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import info.papdt.express.helper.support.Settings
import kotlinx.coroutines.*
import kotlin.coroutines.CoroutineContext

abstract class AbsFragment : Fragment(), CoroutineScope {

	protected var rootView: View? = null
	protected lateinit var settings: Settings

	protected abstract fun getLayoutResId(): Int

	private lateinit var job: Job
	override val coroutineContext: CoroutineContext get() = Dispatchers.Main + job

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		job = Job()

		settings = Settings.getInstance(context!!)
	}

    override fun onDestroy() {
        super.onDestroy()
        job.cancel()
    }

	override fun onCreateView(inflater: LayoutInflater, parent: ViewGroup?, state: Bundle?): View? {
		if (parent != null) {
			rootView = inflater.inflate(getLayoutResId(), parent, false)
		} else {
			rootView = inflater.inflate(getLayoutResId(), null)
		}
		return rootView
	}

    fun ui(block: suspend CoroutineScope.() -> Unit) {
        launch(coroutineContext, block = {
            try {
                block()
            } catch (e: Exception) {
                e.printStackTrace()
                throw e
            }
        })
    }

    fun <T> asyncIO(block: suspend CoroutineScope.() -> T): Deferred<T> {
        return async(Dispatchers.IO, block = {
            try {
                block()
            } catch (e: Exception) {
                e.printStackTrace()
                throw e
            }
        })
    }

}
